### PR TITLE
Wire in WebhookTokenAutenticators support

### PR DESCRIPTION
**Pull request by simo5**
This allows to use an external server for Token authentication.

### CODE REVIEW

**Comment by simo5** [comment]
@mrogers950 @enj PTAL

File: pkg/cmd/server/apis/config/types.go [hunk 1]
```diff
@@ -399,6 +399,9 @@ type MasterConfig struct {
 	EtcdConfig *EtcdConfig
 	// OAuthConfig, if present start the /oauth endpoint in this process
 	OAuthConfig *OAuthConfig
+	// WebhookTokenAuthnConfig, if present configures remote token reviewers
+	WebhookTokenAuthenticators []WebhookTokenAuthenticators
```
Comment (liggitt) [inline pkg/cmd/server/apis/config/types.go]: probably belongs under MasterAuthConfig

Comment (liggitt) [reply pkg/cmd/server/apis/config/types.go]: do we have a use case for allowing more than one, or is that just future-proofing?

File: pkg/cmd/server/apis/config/types.go [hunk 2]
```diff
@@ -828,6 +831,15 @@ type DNSConfig struct {
 	AllowRecursiveQueries bool
 }
 
+type WebhookTokenAuthenticators struct {
```
Comment (liggitt) [inline pkg/cmd/server/apis/config/types.go]: type should be singular if it describes a single authenticator

File: pkg/cmd/server/apis/config/types.go (hunk 1 above)
Comment (simo5) [reply pkg/cmd/server/apis/config/types.go]: Maybe seem that it is more similar to OAuthConfig than RequestHeaderAuthenticationOptions as an auth type, why do we not have OAuthConfig under MasterauthConfig ?

File: pkg/cmd/server/apis/config/types.go (hunk 2 above)
Comment (simo5) [reply pkg/cmd/server/apis/config/types.go]: Ack

File: pkg/cmd/server/kubernetes/master/master_config.go [hunk 3]
```diff
@@ -618,6 +618,7 @@ func defaultOpenAPIConfig(config configapi.MasterConfig) *openapicommon.Config {
 			},
 		}
 	}
+	// TODO: handle WebhookTokenAuthenticators somehow ?
```
Comment (liggitt) [inline pkg/cmd/server/kubernetes/master/master_config.go]: probably just as an additional condition in `if len(config.ServiceAccountConfig.PublicKeyFiles) > 0 || ... token webhooks ... {`

File: pkg/cmd/server/apis/config/types.go (hunk 1 above)
Comment (liggitt) [reply pkg/cmd/server/apis/config/types.go]: OAuth component is distinct from the master API. OAuth config is for the OAuth server. MasterAuthConfig is for the API server.

Comment (liggitt) [reply pkg/cmd/server/apis/config/types.go]: > Maybe seem that it is more similar to OAuthConfig than RequestHeaderAuthenticationOptions as an auth type

RequestHeaderAuthenticationOptions and WebhookTokenAuthenticators both mirror upstream API authentication configuration options. We should group them under MasterAuthConfig.

File: pkg/cmd/server/kubernetes/master/master_config.go (hunk 3 above)
Comment (simo5) [reply pkg/cmd/server/kubernetes/master/master_config.go]: yeah I wasn't sure about that I'll do as you suggest

File: pkg/cmd/server/apis/config/types.go (hunk 1 above)
Comment (simo5) [reply pkg/cmd/server/apis/config/types.go]: Ok, makes sense.

File: pkg/cmd/server/apis/config/v1/types.go [hunk 4]
```diff
@@ -711,6 +714,17 @@ type DNSConfig struct {
...
 }
 
+// WebhookTokenAuthenticators holds the necessary configuation options for
+// external token authenticators
+type WebhookTokenAuthenticator struct {
+	// WebhookTokenAuthnConfigFile is a path to a Kubeconfig file with the webhook configuration
+	WebhookTokenAuthnConfigFile string `json:"webhookTokenAuthnConfigFile"`
+	// WebhookTokenAuthnCacheTTL indicates how long an authentication result should be cached.
+	// It takes a valid time duration string (e.g. "5m").
+	// If empty, you get the default timeout. If zero (e.g. "0m"), caching is disabled
```
Comment (enj) [inline pkg/cmd/server/apis/config/v1/types.go]: Probably should state what the default timeout is (I see no code that does defaulting in this PR).

File: pkg/cmd/server/apis/config/validation/master.go [hunk 5]
```diff
@@ -192,6 +192,24 @@ func ValidateMasterAuthConfig(config configapi.MasterAuthConfig, fldPath *field.
 		validationResults.AddErrors(field.Required(fldPath.Child("requestHeader.extraHeaderPrefixes"), "must be specified for a secure connection"))
 	}
 
+	for _, wta := range config.WebhookTokenAuthenticators {
```
Comment (enj) [inline pkg/cmd/server/apis/config/validation/master.go]: unit test

File: pkg/cmd/server/origin/authenticator.go [hunk 6]
```diff
@@ -108,6 +109,18 @@ func newAuthenticator(config configapi.MasterConfig, accessTokenGetter oauthclie
 			group.NewTokenGroupAdder(oauthTokenAuthenticator, []string{bootstrappolicy.AuthenticatedOAuthGroup}))
 	}
 
+	for _, wta := range config.AuthConfig.WebhookTokenAuthenticators {
```
Comment (enj) [inline pkg/cmd/server/origin/authenticator.go]: integration test (which I know you are already working on)

**Comment by enj** [comment]
> do we have a use case for allowing more than one, or is that just future-proofing?

@liggitt I specifically asked for this from @simo5.  I see no reason not to have it, and I really do not want to try to hack it in later.  It is assumed that all such remote token authenticators are allowed to see all tokens - if you do not trust all of them at that level, then you should not put them in your config.

**Comment by enj** [comment]
@openshift/api-review

File: pkg/cmd/server/apis/config/v1/types.go (hunk 4 above)
Comment (simo5) [reply pkg/cmd/server/apis/config/v1/types.go]: There is currently no default, you are required tio state the CacheTTL

Comment (enj) [reply pkg/cmd/server/apis/config/v1/types.go]: Upstream defaults this to 2 minutes.  @liggitt any preference?

File: test/integration/webhook_token_auth_test.go [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	clientConfig, err := testutil.GetClusterAdminClientConfig(clusterAdminKubeConfig)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	// Ex
```
Comment (mrogers950) [inline test/integration/webhook_token_auth_test.go]: External / Example / or .. ?

File: pkg/cmd/server/apis/config/types.go [hunk 8]
```diff
@@ -399,6 +399,7 @@ type MasterConfig struct {
 	EtcdConfig *EtcdConfig
 	// OAuthConfig, if present start the /oauth endpoint in this process
 	OAuthConfig *OAuthConfig
+
```
Comment (mrogers950) [inline pkg/cmd/server/apis/config/types.go]: unnecessary change

File: pkg/cmd/server/apis/config/v1/types.go [hunk 9]
```diff
@@ -247,6 +247,7 @@ type MasterConfig struct {
 	EtcdConfig *EtcdConfig `json:"etcdConfig"`
 	// OAuthConfig, if present start the /oauth endpoint in this process
 	OAuthConfig *OAuthConfig `json:"oauthConfig"`
+
```
Comment (mrogers950) [inline pkg/cmd/server/apis/config/v1/types.go]: same as above

File: test/integration/webhook_token_auth_test.go (hunk 7 above)
Comment (simo5) [reply test/integration/webhook_token_auth_test.go]: ignore, remnant of my copy/paste from the oidc test

File: pkg/cmd/server/apis/config/v1/types.go (hunk 9 above)
Comment (simo5) [reply pkg/cmd/server/apis/config/v1/types.go]: yeah remamnt of a previous incarnation, will remove, thanks for catching

File: pkg/cmd/server/apis/config/v1/types.go (hunk 4 above)
Comment (simo5) [reply pkg/cmd/server/apis/config/v1/types.go]: I've added a 10s default

File: pkg/cmd/server/apis/config/validation/master.go (hunk 5 above)
Comment (simo5) [reply pkg/cmd/server/apis/config/validation/master.go]: done

**Comment by simo5** [comment]
I think we are ready for final reviews
@mrogers950 @enj @liggitt PTAL

File: pkg/cmd/server/apis/config/validation/master_test.go [hunk 10]
```diff
@@ -525,3 +527,79 @@ func TestValidateIngressIPNetworkCIDR(t *testing.T) {
 		}
 	}
 }
+
+func TestValidateiMasterAuthConfig(t *testing.T) {
```
Comment (mrogers950) [inline pkg/cmd/server/apis/config/validation/master_test.go]: typo: ..Validate**i**Master..

File: pkg/cmd/server/apis/config/v1/conversions.go [hunk 11]
```diff
@@ -136,6 +136,13 @@ func SetDefaults_MasterConfig(obj *MasterConfig) {
 			obj.AdmissionConfig.PluginConfig[pluginName] = &AdmissionPluginConfig{}
 		}
 	}
+
+	// Set defaults Cache TTLs for external Webhook Token Reviewers
+	for _, wta := range obj.AuthConfig.WebhookTokenAuthenticators {
+		if len(wta.CacheTTL) == 0 {
+			wta.CacheTTL = "10s"
```
Comment (enj) [inline pkg/cmd/server/apis/config/v1/conversions.go]: @openshift/api-review any opinion on the default?

File: pkg/cmd/server/apis/config/validation/master_test.go [hunk 12]
```diff
@@ -525,3 +527,79 @@ func TestValidateIngressIPNetworkCIDR(t *testing.T) {
...
+		},
+	}
+	for _, test := range testCases {
+		config := configapi.MasterAuthConfig{
+			RequestHeader:              test.RequestHeader,
+			WebhookTokenAuthenticators: test.WebhookTokenAuthenticators,
+		}
+		errors := ValidateMasterAuthConfig(config, nil)
+		errorCount := len(errors.Errors)
+		if test.errorCount != errorCount {
```
Comment (enj) [inline pkg/cmd/server/apis/config/validation/master_test.go]: Compare the actual errors, not the count.

File: test/integration/webhook_token_auth_test.go [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+
+	// Get master config
+	masterOptions, err := testserver.DefaultMasterOptions()
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	defer testserver.CleanupMasterEtcd(t, masterOptions)
+
+	// Set up a dummy authenticator server
+	authServer := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```
Comment (enj) [inline test/integration/webhook_token_auth_test.go]: This test should also fail if the test server is not called.

File: test/integration/webhook_token_auth_test.go [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+	}
+
+	tokenResponse := kauthn.TokenReview{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "authentication.k8s.io/v1beta1",
+			Kind:       "TokenReview",
+		},
+		Status: kauthn.TokenReviewStatus{
+			Authenticated: true,
+			User: kauthn.UserInfo{
```
Comment (enj) [inline test/integration/webhook_token_auth_test.go]: Send and check all user fields.

File: test/integration/webhook_token_auth_test.go [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+		Spec: kauthn.TokenReviewSpec{Token: authToken},
+	}
+
+	tokenResponse := kauthn.TokenReview{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "authentication.k8s.io/v1beta1",
+			Kind:       "TokenReview",
+		},
+		Status: kauthn.TokenReviewStatus{
+			Authenticated: true,
```
Comment (enj) [inline test/integration/webhook_token_auth_test.go]: Probably want to exercise Authenticated and Error.

File: test/integration/webhook_token_auth_test.go [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+				Cluster:  "authService",
+				AuthInfo: "apiServer",
+			},
+		},
+	}
+	if err := kclientcmd.WriteToFile(authConfigObj, authConfigFile.Name()); err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	masterOptions.AuthConfig.WebhookTokenAuthenticators = []configapi.WebhookTokenAuthenticator{
```
Comment (enj) [inline test/integration/webhook_token_auth_test.go]: Should check with more than one - i.e. have the first one fail and the second on succeed.

File: test/integration/webhook_token_auth_test.go [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+	configapi "github.com/openshift/origin/pkg/cmd/server/apis/config"
+	"github.com/openshift/origin/pkg/oc/cli/cmd"
+	userclient "github.com/openshift/origin/pkg/user/generated/internalclientset/typed/user/internalversion"
+	testutil "github.com/openshift/origin/test/util"
+	testserver "github.com/openshift/origin/test/util/server"
+)
+
+// TestWebhookTokenAuthn checks Tokens directly against an external
+// authenticator
+func TestWebhookTokenAuthn(t *testing.T) {
```
Comment (enj) [inline test/integration/webhook_token_auth_test.go]: @simo5 can you bug KC folks to get us images so the E2E tests can be written ASAP.

File: pkg/cmd/server/apis/config/v1/conversions.go (hunk 11 above)
Comment (enj) [reply pkg/cmd/server/apis/config/v1/conversions.go]: Also, is not this mutating a copy instead of the real object?

Comment (simo5) [reply pkg/cmd/server/apis/config/v1/conversions.go]: Should be fixed now, also changed default to 2m as per upstream

File: pkg/cmd/server/apis/config/validation/master_test.go (hunk 12 above)
Comment (simo5) [reply pkg/cmd/server/apis/config/validation/master_test.go]: done

File: test/integration/webhook_token_auth_test.go (hunk 13 above)
Comment (simo5) [reply test/integration/webhook_token_auth_test.go]: done explicitly, but that was implicit, if the server is not called you do not get the rigth user info

File: test/integration/webhook_token_auth_test.go (hunk 15 above)
Comment (simo5) [reply test/integration/webhook_token_auth_test.go]: I do not think we need test for that here its upstream functionality.
We just want to make sure the webhook is functional

File: test/integration/webhook_token_auth_test.go (hunk 16 above)
Comment (simo5) [reply test/integration/webhook_token_auth_test.go]: hmmm

File: test/integration/webhook_token_auth_test.go (hunk 17 above)
Comment (simo5) [reply test/integration/webhook_token_auth_test.go]: not in the integration test :o)

File: test/integration/webhook_token_auth_test.go (hunk 14 above)
Comment (simo5) [reply test/integration/webhook_token_auth_test.go]: UID and Extras do not carry over to user Info, so I am adding a check froor Groups only.

File: test/integration/webhook_token_auth_test.go (hunk 17 above)
Comment (enj) [reply test/integration/webhook_token_auth_test.go]: I mean send an email :p

File: test/integration/webhook_token_auth_test.go (hunk 14 above)
Comment (enj) [reply test/integration/webhook_token_auth_test.go]: Check scopes by doing a call that will fail due to the scope.

Comment (liggitt) [reply test/integration/webhook_token_auth_test.go]: > UID and Extras do not carry over to user Info

Really? They should, right?

Comment (enj) [reply test/integration/webhook_token_auth_test.go]: > Really? They should, right?

Pretty sure @simo5 meant "UID and Extras do not carry over to `get user ~`"

Comment (simo5) [reply test/integration/webhook_token_auth_test.go]: what I mean is the WhoAmI call returns a structure without the UID and there are no Extra fields at all.
@enj please suggest a call and a scope that would satisfy your idea of a check.

**Comment by simo5** [comment]
@deads2k can you please API Review this ?

File: pkg/cmd/server/apis/config/helpers.go [hunk 18]
```diff
@@ -238,6 +238,10 @@ func GetMasterFileReferences(config *MasterConfig) []*string {
 		refs = append(refs, &config.AuthConfig.RequestHeader.ClientCA)
 	}
 
+	for _, wta := range config.AuthConfig.WebhookTokenAuthenticators {
```
Comment (deads2k) [inline pkg/cmd/server/apis/config/helpers.go]: Can you use the index instead?  I have trouble keeping my head straight on when copies are made and the originals are used.  The rest of this method uses `for k := range foo; append(&foo[k])`

File: pkg/cmd/server/apis/config/v1/conversions.go [hunk 19]
```diff
@@ -136,6 +136,13 @@ func SetDefaults_MasterConfig(obj *MasterConfig) {
 			obj.AdmissionConfig.PluginConfig[pluginName] = &AdmissionPluginConfig{}
 		}
 	}
+
+	// Set defaults Cache TTLs for external Webhook Token Reviewers
+	for i := range obj.AuthConfig.WebhookTokenAuthenticators {
+		if len(obj.AuthConfig.WebhookTokenAuthenticators[i].CacheTTL) == 0 {
+			obj.AuthConfig.WebhookTokenAuthenticators[i].CacheTTL = "2m"
```
Comment (deads2k) [inline pkg/cmd/server/apis/config/v1/conversions.go]: The node defaults to 5m.  Any reason for a different default?  If this was already litigated, I don't feel strongly.

File: pkg/cmd/server/apis/config/helpers.go (hunk 18 above)
Comment (simo5) [reply pkg/cmd/server/apis/config/helpers.go]: ok

File: pkg/cmd/server/apis/config/v1/conversions.go (hunk 19 above)
Comment (simo5) [reply pkg/cmd/server/apis/config/v1/conversions.go]: upstream defauls to 2m for CacheTTL see other comments made previously

Comment (deads2k) [reply pkg/cmd/server/apis/config/v1/conversions.go]: > upstream defauls to 2m for CacheTTL see other comments made previously

Ok.

**Comment by deads2k** [comment]
api approved.

File: pkg/cmd/server/apis/config/serialization_test.go [hunk 20]
```diff
@@ -215,6 +215,15 @@ func fuzzInternalObject(t *testing.T, forVersion schema.GroupVersion, item runti
 				}
 			}
 
+			for i := range obj.AuthConfig.WebhookTokenAuthenticators {
+				if len(obj.AuthConfig.WebhookTokenAuthenticators[i].ConfigFile) == 0 {
+					continue
```
[the remark this answers is not in the document]
Comment (simo5) [reply pkg/cmd/server/apis/config/serialization_test.go]: Does not default the TTL if there is no file to start with

[the remark this answers is not in the document]
Comment (simo5) [reply pkg/cmd/server/apis/config/serialization_test.go]: (config is already wrong)

File: test/integration/webhook_token_auth_test.go [hunk 21]
```diff
@@ -0,0 +1,241 @@
...
+	authServerWasCalled := false
+	authToken := "Anything-goes!"
+	authTestUser := "testuser"
+	authTestUID := "42"
+	authTestGroups := []string{"testgroup"}
```
[the remark this answers is not in the document]
Comment (simo5) [reply test/integration/webhook_token_auth_test.go]: no need

File: test/integration/webhook_token_auth_test.go [hunk 22]
```diff
@@ -0,0 +1,241 @@
...
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	defer os.Remove(caFile.Name())
+	if err := ioutil.WriteFile(caFile.Name(), authLocalhostCert, os.FileMode(0600)); err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	// Get master config
+	masterOptions, err := testserver.DefaultMasterOptions()
```
[the remark this answers is not in the document]
Comment (simo5) [reply test/integration/webhook_token_auth_test.go]: ?

File: test/integration/webhook_token_auth_test.go [hunk 23]
```diff
@@ -0,0 +1,241 @@
...
+
+	clientConfig, err := testutil.GetClusterAdminClientConfig(clusterAdminKubeConfig)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	// Try to authenticate with a token that can be validated only by our
+	// external token reviewer
+	userConfig := &restclient.Config{
+		Host: clientConfig.Host,
```
[the remark this answers is not in the document]
Comment (simo5) [reply test/integration/webhook_token_auth_test.go]: ?

File: test/integration/webhook_token_auth_test.go [hunk 24]
```diff
@@ -0,0 +1,241 @@
...
+		BearerToken: authToken,
+	}
+	userClient, err := userclient.NewForConfig(userConfig)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	userWhoamiOptions := cmd.WhoAmIOptions{UserInterface: userClient.Users(), Out: ioutil.Discard}
+	retrievedUser, err := userWhoamiOptions.WhoAmI()
+	if err != nil {
+		t.Errorf("unexpected error: %v", err)
```
[the remark this answers is not in the document]
Comment (simo5) [reply test/integration/webhook_token_auth_test.go]: Not really needed, other tests do the same

File: test/integration/webhook_token_auth_test.go [hunk 25]
```diff
@@ -0,0 +1,241 @@
...
+		},
+	}
+	if err := kclientcmd.WriteToFile(authConfigObj, authConfigFile.Name()); err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	masterOptions.AuthConfig.WebhookTokenAuthenticators = []configapi.WebhookTokenAuthenticator{
+		{
+			ConfigFile: authConfigFile.Name(),
+			CacheTTL:   "10s",
```
[the remark this answers is not in the document]
Comment (simo5) [reply test/integration/webhook_token_auth_test.go]: defaulting is already checked in uint tests, no need to check it here

File: test/integration/webhook_token_auth_test.go [hunk 26]
```diff
@@ -0,0 +1,241 @@
...
+	if !authServerWasCalled {
+		t.Errorf("Server was not called in the test")
+	}
+
+	oauthClient, err := oauthclient.NewForConfig(userConfig)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	err = oauthClient.Oauth().OAuthAccessTokens().Delete("XYZ", &metav1.DeleteOptions{})
+	if !kerrors.IsForbidden(err) {
```
[the remark this answers is not in the document]
Comment (simo5) [reply test/integration/webhook_token_auth_test.go]: Why ?

[the remark this answers is not in the document]
Comment (enj) [reply test/integration/webhook_token_auth_test.go]: Because we should know we got forbidden via scope and not "the token stopped working for random reason"

File: pkg/cmd/server/apis/config/validation/master_test.go [hunk 27]
```diff
@@ -525,3 +527,86 @@ func TestValidateIngressIPNetworkCIDR(t *testing.T) {
...
+		},
+	}
+	for _, test := range testCases {
+		config := configapi.MasterAuthConfig{
+			RequestHeader:              test.RequestHeader,
+			WebhookTokenAuthenticators: test.WebhookTokenAuthenticators,
+		}
+		errors := ValidateMasterAuthConfig(config, nil)
+		if len(test.expectedErrors) != len(errors.Errors) {
+			t.Errorf("%s: expected %v errors, got %v", test.testName, test.expectedErrors, errors.Errors)
```
[the remark this answers is not in the document]
Comment (simo5) [reply pkg/cmd/server/apis/config/validation/master_test.go]: Ah, true, but then, the test already failed :-)
Will fix

File: test/integration/webhook_token_auth_test.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (enj) [reply test/integration/webhook_token_auth_test.go]: Hm I guess serialization tests may be enough.

File: test/integration/webhook_token_auth_test.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (enj) [reply test/integration/webhook_token_auth_test.go]: Just because other people write their tests incorrectly is not a valid reason to continue doing the wrong thing.

File: test/integration/webhook_token_auth_test.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (enj) [reply test/integration/webhook_token_auth_test.go]:
```
userConfig := AnonymousClientConfig(...)
userConfig.BearerToken = ...
```

File: pkg/cmd/server/apis/config/serialization_test.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (enj) [reply pkg/cmd/server/apis/config/serialization_test.go]: Sure, but then why do we not want to default?  It will just make two errors show up and make the user think that both are required.

File: test/integration/webhook_token_auth_test.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (enj) [reply test/integration/webhook_token_auth_test.go]: Just cosmetic for increasing readability since you do not need that variable until later.

File: test/integration/webhook_token_auth_test.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (enj) [reply test/integration/webhook_token_auth_test.go]: OK

File: pkg/cmd/server/apis/config/serialization_test.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (simo5) [reply pkg/cmd/server/apis/config/serialization_test.go]: ok removing both

File: test/integration/webhook_token_auth_test.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (simo5) [reply test/integration/webhook_token_auth_test.go]: ok

File: test/integration/webhook_token_auth_test.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (simo5) [reply test/integration/webhook_token_auth_test.go]: ooook </lazy>

File: test/integration/webhook_token_auth_test.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (simo5) [reply test/integration/webhook_token_auth_test.go]: DEfaulting happens before we override this part of the config anyway, so it wouldn't work :-)

File: pkg/cmd/server/origin/authenticator.go [hunk 28]
```diff
@@ -108,6 +109,18 @@ func newAuthenticator(config configapi.MasterConfig, accessTokenGetter oauthclie
...
+		ttl, err := time.ParseDuration(wta.CacheTTL)
+		if err != nil {
+			return nil, nil, fmt.Errorf("Error converting CacheTTL='%s' to duration", wta.CacheTTL)
+		}
+		webhookTokenAuthenticator, err := webhooktoken.New(wta.ConfigFile, ttl)
+		if err != nil {
+			return nil, nil, fmt.Errorf("Failed to create webhook token authenticator for ConfigFile='%s'", wta.ConfigFile)
```
Comment (enj) [inline pkg/cmd/server/origin/authenticator.go]: Include the error that caused the failure:

`return nil, nil, fmt.Errorf("Failed to create webhook token authenticator for ConfigFile=%q: %v", wta.ConfigFile, err)`

Also replace `'%s'` with `%q` on both messages.

**Comment by simo5** [comment]
Flake: #18982
